tests/test_jax_nonbonded.py

```diff
@@ -0,0 +1,151 @@
+
+from functools import partial
+from typing import Tuple, Callable
+
+Conf = Params = Box = ChargeMask = LJMask = LambdaPlaneIdxs = LambdaOffsetIdxs = np.array
+Lamb = Beta = Cutoff = Energy = float
+
+nonbonded_args = Conf, Params, Box, Lamb, ChargeMask, LJMask, Beta, Cutoff, LambdaPlaneIdxs, LambdaOffsetIdxs
+NonbondedArgs = Tuple[nonbonded_args]
+NonbondedFxn = Callable[[*nonbonded_args], Energy]
+
+
+def generate_random_inputs(n_atoms: int, dim: int = 3) -> NonbondedArgs:
+    # jittered
+    offsets = onp.array([onp.arange(n_atoms)] * dim).T
+    assert (offsets.shape == (n_atoms, dim))
+    for i in range(dim):
+        onp.random.shuffle(offsets[:, i])
+    conf = (0.5 * randn(n_atoms, dim) + offsets)
+    conf += conf.min()
+    params = rand(n_atoms, 3)
+    params[:, 0] -= np.mean(params[:, 0])
+
+    box = np.diag(0.5 + 0.5 * rand(3)) * 10 * n_atoms
+
+    lamb = rand()
+
+    charge_rescale_mask = onp.ones((n_atoms, n_atoms))
+    lj_rescale_mask = onp.ones((n_atoms, n_atoms))
+    for _ in range(n_atoms):
+        i, j = randint(n_atoms, size=2)
+        charge_rescale_mask[i, j] = 0.0
+        lj_rescale_mask[i, j] = 0.0
+
+    beta = rand() + 1
+    cutoff = rand() + 0.5
+
+    lambda_plane_idxs = randint(low=-2, high=2, size=n_atoms)
+    lambda_offset_idxs = randint(low=-2, high=2, size=n_atoms)
+
+    args = conf, params, box, lamb, charge_rescale_mask, lj_rescale_mask, beta, cutoff, lambda_plane_idxs, lambda_offset_idxs
+
+    return args
+
+
+def compare_two_potentials(u_a: NonbondedFxn, u_b: NonbondedFxn, args: NonbondedArgs, differentiate_wrt=(0, 1, 3)):
+    """Assert that energies and derivatives w.r.t. request argnums are close"""
+    value_and_grads = partial(value_and_grad, argnums=differentiate_wrt)
+    energy_a, gradients_a = value_and_grads(u_a)(*args)
+    energy_b, gradients_b = value_and_grads(u_b)(*args)
+
+    assert energy_a == energy_b
+    onp.testing.assert_almost_equal(energy_a, energy_b)
+    for (g_a, g_b) in zip(gradients_a, gradients_b):
+        onp.testing.assert_allclose(g_a, g_b)
+
+
+def _nonbonded_v3_clone(
+        conf,
+        params,
+        box,
+        lamb,
+        charge_rescale_mask,
+        lj_rescale_mask,
+        beta,
+        cutoff,
+        lambda_plane_idxs,
+        lambda_offset_idxs,
+        runtime_validate=False,
+):
+    """See docstring of nonbonded_v3 for more details
+
+    This is here just for testing purposes...
+    """
+
+    N = conf.shape[0]
+
+    if conf.shape[-1] == 3:
+        conf = convert_to_4d(conf, lamb, lambda_plane_idxs, lambda_offset_idxs, cutoff)
+
+    # make 4th dimension of box large enough so its roughly aperiodic
+    if box is not None:
+        box_4d = np.eye(4) * 1000
+        box_4d = index_update(box_4d, index[:3, :3], box)
+    box = box_4d
+
+    # TODO: break this into more manageable blocks
+    inds_i, inds_j = get_all_pairs_indices(N)
+    # n_interactions = len(inds_i)
```

Comment:
extra comment?

Reply:
Consolidated in https://github.com/proteneer/timemachine/pull/453/commits/fcf2a98def1bd877cd8f310a495ecb31decd5099